File: golang/query/qwordBuilder.go
package query

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	z "github.com/nutzam/zgo"
	"io"
	"regexp"
	"strconv"
	"strings"
)

type QWBuilder struct {
	GOr          string       `json:"gOr"`
	GAnd         string       `json:"gAnd"`
	SepOr        []string     `json:"sepOr"`
	SepAnd       []string     `json:"sepAnd"`
	QuoteBegin   []string     `json:"quoteBegin"`
	QuoteEnd     []string     `json:"quoteEnd"`
	BracketBegin []string     `json:"bracketBegin"`
	BracketEnd   []string     `json:"bracketEnd"`
	Rules        []*QWordRule `json:"rules"`
}

// 返回一个qwordBuilder的描述信息
func (qb *QWBuilder) String() string {
	sb := z.StringBuilder()
	kwidth := 15
	sb.Append("QWorldBuilder Setup").EOL()
	sb.Append(z.AlignLeft("gOr", kwidth, ' ')).Append(": ").Append(qb.GOr).EOL()
	sb.Append(z.AlignLeft("gAnd", kwidth, ' ')).Append(": ").Append(qb.GAnd).EOL()
	sb.Append(z.AlignLeft("sepOr", kwidth, ' ')).Append(": ").Append(qb.SepOr).EOL()
	sb.Append(z.AlignLeft("sepAnd", kwidth, ' ')).Append(": ").Append(qb.SepAnd).EOL()
	sb.Append(z.AlignLeft("quoteBegin", kwidth, ' ')).Append(": ").Append(qb.QuoteBegin).EOL()
	sb.Append(z.AlignLeft("quoteEnd", kwidth, ' ')).Append(": ").Append(qb.QuoteEnd).EOL()
	sb.Append(z.AlignLeft("bracketBegin", kwidth, ' ')).Append(": ").Append(qb.BracketBegin).EOL()
	sb.Append(z.AlignLeft("bracketEnd", kwidth, ' ')).Append(": ").Append(qb.BracketEnd).EOL()
	// QWordRule
	if len(qb.Rules) > 0 {
		sb.Append(z.AlignLeft("rules", kwidth, ' ')).Append(": ").EOL()
		for i := 0; i < len(qb.Rules); i++ {
			rule := qb.Rules[i]
			sb.Append(z.DupChar(' ', kwidth)).Append(fmt.Sprintf(" %2d. {", i))
			sb.Append(z.DupChar(' ', kwidth)).Append(z.DupChar(' ', kwidth)).Append(fmt.Sprint("{")).EOL()
			sb.Append(z.DupChar(' ', kwidth+9)).Append(fmt.Sprintf("key    : %s", rule.Key)).EOL()
			sb.Append(z.DupChar(' ', kwidth+9)).Append(fmt.Sprintf("regex  : %s", rule.Regex.String())).EOL()
			sb.Append(z.DupChar(' ', kwidth+9)).Append(fmt.Sprintf("seg    : %s", rule.Seg)).EOL()
			sb.Append(z.DupChar(' ', kwidth+9)).Append(fmt.Sprintf("type   : %v", rule.Type)).EOL()
			sb.Append(z.DupChar(' ', kwidth+5)).Append(fmt.Sprint("}")).EOL()
		}
	}
	return sb.String()
}

// 返回一个使用默认参数的qwordBuilder
func QWordBuilder(rule ...io.Reader) *QWBuilder {
	qb := &QWBuilder{
		"OR",
		"AND",
		[]string{" "},
		[]string{","},
		[]string{"\"", "'"},
		[]string{"\"", "'"},
		[]string{"{", "[", "("},
		[]string{"}", "]", ")"},
		[]*QWordRule{},
	}
	if len(rule) > 0 {
		// 加载rule
		qb.LoadRules(rule[0])
	}
	return qb
}

// 设置qwordBuilder的参数, 使用json格式字符串
// 仅仅可以修改 gOr, gAnd, sepOr, sepAnd 四个参数
func (qb *QWBuilder) Setup(sjson string) *QWBuilder {
	setup := new(QWBuilder)
	err := json.Unmarshal([]byte(sjson), setup)
	if err != nil {
		panic(err)
	}
	qb.GOr = z.SBlank(setup.GOr, qb.GOr)
	qb.GAnd = z.SBlank(setup.GAnd, qb.GAnd)
	if setup.SepOr != nil {
		qb.SepOr = setup.SepOr
	}
	if setup.SepAnd != nil {
		qb.SepAnd = setup.SepAnd
	}
	// TODO 需要check一下设置的项
	return qb
}

// 读取解析规则字符串
func (qb *QWBuilder) LoadRulesStr(str string) *QWBuilder {
	return qb.LoadRules(strings.NewReader(str))
}

// 读取解析规则, 传入一个实现了io.Reader的对象即可
func (qb *QWBuilder) LoadRules(reader io.Reader) *QWBuilder {
	bufreader := bufio.NewReader(reader)
	var line string
	var lnum int
	var err error
	var rules []*QWordRule = make([]*QWordRule, 0)
	for true {
		line, err = bufreader.ReadString('\n')
		if err == io.EOF {
			err = nil
			break
		} else if err != nil {
			break
		}
		lnum++
		z.DebugPrintf("rule %2d : %s", lnum, line)
		// 忽略空行与注释行
		line = strings.TrimSpace(line)
		if z.IsBlank(line) || strings.HasPrefix(line, "#") {
			continue
		}
		// 解析行信息
		if strings.HasPrefix(line, "$") {
			qwRule := new(QWordRule)
			// 获得key
			var fpos int = strings.Index(line, ":")
			if fpos < 0 {
				err = errors.New(fmt.Sprintf("invalid rule line %d : %s", lnum, line))
				break
			}
			qwRule.Key = strings.TrimSpace(line[1:fpos])

			// 获得regex
			var expr string
			if line[fpos+1] == ':' {
				// 简要模式
				expr = "^(" + strings.TrimSpace(line[fpos+2:]) + ")(.*)$"
			} else {
				// 普通模式
				expr = strings.TrimSpace(line[fpos+1:])
			}
			regex, regerr := regexp.Compile(expr)
			if regerr != nil {
				err = regerr
				break
			}
			qwRule.Regex = regex

			// 读取下一行,必须存在哟
			line, err = bufreader.ReadString('\n')
			if err == io.EOF {
				err = errors.New(fmt.Sprintf("rule line(%d) miss next line", lnum))
				break
			} else if err != nil {
				break
			}
			lnum++
			z.DebugPrintf("rule %2d : %s", lnum, line)
			line = strings.TrimSpace(line)
			spos := strings.Index(line, "=")
			if z.IsBlank(line) || !strings.HasPrefix(line, "$") || spos < 0 {
				err = errors.New(fmt.Sprintf("invalid rule line %d : %s", lnum, line))
				break
			}
			// 获得seg
			qwRule.Seg = z.Trim(line[:spos])
			// 获得type
			qcType, qcterr := QCType(strings.TrimSpace(line[spos+1:]))
			if qcterr != nil {
				err = qcterr
				break
			}
			qwRule.Type = qcType
			// 加入到rules
			rules = append(rules, qwRule)
		}
	}
	// 是否有错误
	if err != nil {
		panic(err)
	}
	qb.Rules = rules
	return qb
}

// 解析查询字符串
func (qb *QWBuilder) Parse(kwd string) *QWord {
	kwd = z.TrimExtraSpace(kwd)
	if len(kwd) == 0 {
		return nil
	}
	isGOr, isGAnd, flds, seps := qb.extractFldsAndSeps(kwd)
	// 解析为QWord
	qword := new(QWord)
	qword.CndMap = map[string]*QCnd{}
	sseps := len(seps)
	for i, fld := range flds {
		// qcnd
		qc := qb.evalQCnd(fld)
		if qc != nil {
			qword.Cnds = append(qword.Cnds, qc)
			qword.CndMap[qc.Key] = qc
			// 分隔符
			if i < sseps {
				sp := "&"
				if isGOr {
					sp = "|"
				} else if isGAnd {
					sp = "&"
				} else if z.IsInStrings(qb.SepOr, string(seps[i])) {
					sp = "|"
				} else if z.IsInStrings(qb.SepAnd, string(seps[i])) {
					sp = "&"
				}
				qword.Rels = append(qword.Rels, sp)
			}
		} else {
			qword.Unmatchs = append(qword.Unmatchs, fld)
		}
	}
	return qword
}

// 配备并重新组合查询语句, 给出QCnd对象
func (qb *QWBuilder) evalQCnd(fld string) *QCnd {
	for _, rule := range qb.Rules {
		if rule.Regex.MatchString(fld) {
			z.DebugPrintf("fld [%s] match regex [%s]\n", fld, rule.Regex.String())
			groups := rule.Regex.FindStringSubmatch(fld)
			for _, grp := range groups {
				z.DebugPrintf("    g_%s\n", grp)
			}
			extractStr := z.Trim(findMatchStr(rule.Seg, groups))
			// 生成QCnd
			qc := new(QCnd)
			qc.Key = rule.Key
			qc.Origin = fld
			qc.Plain = extractStr
			qc.Type = rule.Type
			switch qc.Type {
			case String:
				qc.Value = extractStr
			case Regex:
				qc.Value = regexp.MustCompile(extractStr)
			case IntRegion, LongRegion, DateRegion:
				qc.Value = z.MakeRegion(extractStr)
			case StringEnum:
				senum := extractStr[1 : len(extractStr)-1]
				qc.Value = z.SplitIgnoreBlank(senum, ",")
			case IntEnum:
				ienum := extractStr[1 : len(extractStr)-1]
				iarray := make([]int, 0, 5)
				for _, ie := range z.SplitIgnoreBlank(ienum, ",") {
					ione, _ := strconv.Atoi(ie)
					iarray = append(iarray, ione)
				}
				qc.Value = iarray
			case Json:
				jmap := new(map[string]interface{})
				jerr := z.JsonFromString(extractStr, jmap)
				if jerr != nil {
					qc.Value = jmap
				}
			default:
				// unsupport
				qc.Value = nil
			}
			return qc
		}
	}
	z.DebugPrintf("fld [%s] miss match\n", fld)
	return nil
}

// 简单的查找出对应的group
func findMatchStr(seg string, groups []string) string {
	// FIXME 这里先简单实现下, 默认格式都是 '#{x}'
	seg = seg[2 : len(seg)-1]
	pos, err := strconv.Atoi(seg)
	if err != nil {
		panic(err)
	}
	return groups[pos]
}

// 提取查询语句与连接符
func (qb *QWBuilder) extractFldsAndSeps(kwd string) (isGOr, isGAnd bool, flds []string, seps []byte) {
	// 判断全局参数 (先强制认为 'xxx:')
	// FIXME 后面改成正则匹配, 可以多加空格
	isGOr = strings.HasPrefix(kwd, qb.GOr+":")
	isGAnd = strings.HasPrefix(kwd, qb.GAnd+":")
	if isGOr || isGAnd {
		kwd = kwd[strings.Index(kwd, ":")+1:]
	}
	// 分割查询字符串
	flds = make([]string, 0, 5)
	seps = make([]byte, 0, 5)
	// FIXME 这个地方需要在好好想想
	// 保证一个特殊情况下的正确性, 比如或是" "与是"," kwd中有类似"xxx , yyy"这样的语句,需要把处理
	// 这个特例是一搬在" "当做或的情况下
	if z.IsInStrings(qb.SepOr, " ") {
		for _, sAnd := range qb.SepAnd {
			ss := " " + sAnd + " "
			kwd = strings.Replace(kwd, ss, sAnd, -1)
		}
	}
	// 开始解析
	kszie := len(kwd)
	fld := z.StringBuilder()
	for i := 0; i < kszie; i++ {
		c := kwd[i]
		z.DebugPrintf("extract kwd %3d -> [%s]\n", i, string(c))
		// 如果是引用
		if pos := z.IndexOfStrings(qb.QuoteBegin, string(c)); pos > 0 {
			// 这里面的空格啥的就不会被跳过了, 会一直取到引用结束, 逃逸字符会调整一下
			for i++; i < kszie; i++ {
				c = kwd[i]
				if string(c) == qb.QuoteEnd[pos] {
					break
				} else if c == '\\' {
					// FIXME 这里没有做防守
					i++
					fld.Append(kwd[i])
				} else {
					fld.Append(c)
				}
			}
			continue
		}
		// 如果是括弧
		if z.IsInStrings(qb.BracketBegin, string(c)) {
			fld.Append(c)
			for i++; i < kszie; i++ {
				c = kwd[i]
				fld.Append(c)
				if z.IsInStrings(qb.BracketEnd, string(c)) {
					break
				}
			}
			continue
		}
		// 如果是分隔符
		if z.IsInStrings(qb.SepAnd, string(c)) || z.IsInStrings(qb.SepOr, string(c)) {
			cfld := z.Trim(fld.String())
			if len(cfld) > 0 {
				flds = append(flds, cfld)
				seps = append(seps, c)
				fld = z.StringBuilder()
			}
			continue
		}
		// 没什么特殊情况,那就加入到sb
		fld.Append(c)
	}
	// 加入最后一个条件
	lfld := z.Trim(fld.String())
	if len(lfld) > 0 {
		flds = append(flds, lfld)
	}
	// 打印一下
	if z.IsDebugOn() && len(flds) > 0 {
		z.DebugPrintf("kwd : %s\n", kwd)
		for _, f := range flds {
			z.DebugPrintf("fld : %s\n", f)
		}
		for _, s := range seps {
			z.DebugPrintf("sep : '%s'\n", string(s))
		}
	}
	return isGOr, isGAnd, flds, seps
}

File: java/query/src/org/woods/query/QWordBuilder.java
package org.woods.query;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.Reader;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.nutz.lang.Lang;
import org.nutz.lang.Nums;
import org.nutz.lang.Streams;
import org.nutz.lang.Strings;
import org.nutz.lang.segment.Segments;
import org.nutz.lang.util.Context;
import org.nutz.lang.util.Region;

public class QWordBuilder {

    private String gOr;
    private String gAnd;

    private char[] sepOr;
    private char[] sepAnd;

    private char[] quoteBegin;
    private char[] quoteEnd;

    private char[] bracketBegin;
    private char[] bracketEnd;

    private List<QWordRule> rules;

    public QWordBuilder(File f) {
        this(Streams.fileInr(f));
    }

    public QWordBuilder(Reader r) {
        this();
        this.loadRules(r);
    }

    public QWordBuilder() {
        gOr = "OR";
        gAnd = "AND";
        sepOr = new char[]{' ', '\t', '|', '\n'};
        sepAnd = new char[]{',', '&'};
        quoteBegin = new char[]{'"', '\''};
        quoteEnd = new char[]{'"', '\''};
        bracketBegin = new char[]{'[', '(', '{'};
        bracketEnd = new char[]{']', ')', '}'};

        rules = new ArrayList<QWordRule>();
    }

    public QWordBuilder setup(String json) {
        return setup(Lang.map(json));
    }

    public QWordBuilder setup(Map<String, Object> map) {
        gOr = Strings.sBlank(map.get("gOr"), gOr);
        gAnd = Strings.sBlank(map.get("gAnd"), gAnd);
        sepOr = Strings.sBlank(map.get("sepOr"), new String(sepOr)).toCharArray();
        sepAnd = Strings.sBlank(map.get("sepAnd"), new String(sepAnd)).toCharArray();
        return this;
    }

    public QWordBuilder loadRulesStr(String rule) {
        return this.loadRules(new StringReader(rule));
    }

    public QWordBuilder loadRules(Reader r) {
        BufferedReader br = Streams.buffr(r);

        // 按行读取
        try {
            String line;
            int lineNumber = 0;
            while (null != (line = br.readLine())) {
                lineNumber++;
                line = Strings.trim(line);
                // 忽略空行和注释行
                if (Strings.isEmpty(line) || line.startsWith("#"))
                    continue;

                // 如果行以 $ 开头表示一个约束描述
                if (line.startsWith("$")) {
                    QWordRule qr = new QWordRule();
                    // 寻找到第一个 :
                    int pos = line.indexOf(':');
                    // 神码？竟敢木油冒号？！ 抛错对付你 >:D
                    if (pos == -1)
                        throw Lang.makeThrow("invalid rule line %d : %s", lineNumber, line);
                    // 获取名称
                    qr.key = Strings.trim(line.substring(1, pos));
                    String regex;

                    // 看看是否是简要模式
                    if (line.charAt(pos + 1) == ':') {
                        regex = "^(" + Strings.trim(line.substring(pos + 2)) + ")(.*)$";
                    }
                    // 普通模式
                    else {
                        regex = Strings.trim(line.substring(pos + 1));
                    }
                    // 得到正则表达式
                    qr.regex = Pattern.compile(regex);

                    // 再读一行作为详细描述
                    line = Strings.trim(br.readLine());
                    lineNumber++;

                    // 分隔
                    pos = line.lastIndexOf('=');

                    // 神码？竟敢木油等号？！ 抛错对付你 >:D
                    if (pos == -1)
                        throw Lang.makeThrow("invalid rule line %d : %s", lineNumber, line);

                    qr.type = QCndType.valueOf(Strings.trim(line.substring(pos + 1)));
                    qr.seg = Segments.create(Strings.trim(line.substring(0, pos)));

                    rules.add(qr);

                }
                // 其他的忽略

            }
        }
        catch (IOException e) {
            throw Lang.wrapThrow(e);
        }

        return this;
    }

    public QWord parse(String kwd) {
        if (Strings.isBlank(kwd))
            return null;

        // 建立返回值
        QWord w = new QWord();

        // 判断是否全局为 OR
        Boolean gIsOr = null;
        Boolean gIsAnd = null;
        if (null != gOr && kwd.startsWith(gOr + ":")) {
            kwd = kwd.substring(gOr.length() + 1);
            gIsOr = true;
            w.setDefaultAllAnd(false);
        }
        if (null != gAnd && kwd.startsWith(gAnd + ":")) {
            kwd = kwd.substring(gAnd.length() + 1);
            gIsAnd = true;
            w.setDefaultAllAnd(true);
        }

        // 拆分字符串
        List<String> flds = new ArrayList<String>();
        List<Boolean> seps = new ArrayList<Boolean>();

        _kwd_to_flds(kwd, flds, seps);

        // 准备解析
        Context c = Lang.context();
        int i = 0;
        for (; i < seps.size(); i++) {
            String fld = flds.get(i);
            QCnd cnd = _eval_cnd(c, fld);

            // 加入到关键字中
            if (null != cnd) {
                boolean nextIsAnd = seps.get(i);
                // AND:xxxxx
                if (null != gIsAnd && gIsAnd.booleanValue()) {
                    nextIsAnd = true;
                }
                // OR:xxxx
                else if (null != gIsOr && gIsOr.booleanValue()) {
                    nextIsAnd = false;
                }

                w.add(cnd, nextIsAnd);
            }
        }

        // 处理最后一个
        w.done(_eval_cnd(c, flds.get(i)));

        return w;
    }

    private QCnd _eval_cnd(Context c, String fld) {
        // 首先匹配各个规则
        for (QWordRule rule : rules) {
            Matcher m = rule.regex.matcher(fld);
            if (m.find()) {
                // 收集各个组
                c.clear();
                for (int x = 0; x <= m.groupCount(); x++) {
                    c.set("" + x, m.group(x));
                }

                // 得到抽出字符串
                String str = rule.seg.render(c).toString();

                // FIXME, 空字符串也是一种信息哟
                // 抽出空字符串，那么就表示这个规则啥都木有
                // if (Strings.isBlank(str))
                // return null;

                // 设置与约束
                QCnd cnd = new QCnd();
                cnd.setKey(rule.key);
                cnd.setOrigin(fld);
                cnd.setPlain(str);
                cnd.setType(rule.type);

                // 生成值
                switch (rule.type) {
                case INT:
                    return cnd.setValue(Integer.valueOf(str));
                case IntRegion:
                    Region<Integer> rI = Region.Int(str);
                    return rI.isNull() ? null : cnd.setValue(rI);
                case LongRegion:
                    Region<Long> rL = Region.Long(str);
                    return rL.isNull() ? null : cnd.setValue(rL);
                case DateRegion:
                    Region<Date> rD = Region.Date(str);
                    return rD.isNull() ? null : cnd.setValue(rD);
                case StringEnum:
                    String[] ss = Strings.splitIgnoreBlank(str);
                    return ss == null || ss.length == 0 ? null : cnd.setValue(ss);
                case IntEnum:
                    int[] ii = Nums.splitInt(str);
                    return ii == null || ii.length == 0 ? null : cnd.setValue(ii);
                case Regex:
                    try {
                        return cnd.setValue(Pattern.compile(str));
                    }
                    catch (Exception e) {
                        return null;
                    }
                case Json:
                    try {
                        return cnd.setValue(Lang.map(str));
                    }
                    catch (Exception e) {
                        return null;
                    }

                default:
                    return cnd.setType(QCndType.String).setValue(str);
                }
            }
        }
        return null;
    }

    private void _kwd_to_flds(String kwd, List<String> flds, List<Boolean> seps) {

        char[] cs = kwd.toCharArray();
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < cs.length; i++) {
            char c = cs[i];
            // 如果是 AND 的连接符
            if (Nums.isin(sepAnd, c)) {
                if (!Strings.isBlank(sb)) {
                    flds.add(sb.toString());
                    seps.add(Boolean.TRUE);
                    sb = new StringBuilder();
                }
                // 如果已经有字段了，那么将最后一个连接符设为 "AND"
                else if (!flds.isEmpty()) {
                    seps.set(flds.size() - 1, Boolean.TRUE);
                }
                continue;
            }
            // 如果是 OR 的连接符
            if (Nums.isin(sepOr, c)) {
                if (!Strings.isBlank(sb)) {
                    flds.add(sb.toString());
                    seps.add(Boolean.FALSE);
                    sb = new StringBuilder();
                }
                continue;
            }
            // 空白字符，忽略
            if (Character.isWhitespace(c)) {
                continue;
            }
            // 如果是字符串或者开始，一直读到字符串的结尾
            int pos;
            if ((pos = Nums.indexOf(quoteBegin, c)) >= 0) {
                for (i++; i < cs.length; i++) {
                    c = cs[i];
                    // 读完了字符串
                    if (quoteEnd[pos] == c) {
                        break;
                    }
                    // 逃逸字符
                    else if (c == '\\') {
                        i++;
                        sb.append(cs[i]);
                    }
                    // 其他字符统统读入
                    else {
                        sb.append(c);
                    }
                }
                continue;
            }
            // 如果开始括弧，一直读到括弧结束
            if ((pos = Nums.indexOf(bracketBegin, c)) >= 0) {
                sb.append(c);
                for (i++; i < cs.length; i++) {
                    c = cs[i];
                    sb.append(c);
                    // 读完了括弧
                    if (bracketEnd[pos] == c) {
                        break;
                    }
                }
                continue;
            }
            // 其他的字符，就是普通的增加就是
            sb.append(c);
        }
        // 确保最后一个字段加入列表
        if (sb.length() > 0) {
            flds.add(Strings.trim(sb.toString()));
        }

    }

}
